diff --git a/third_party/lcms/src/cmsplugin.c b/third_party/lcms/src/cmsplugin.c
index 79b145d2d..29ddd9a49 100644
--- a/third_party/lcms/src/cmsplugin.c
+++ b/third_party/lcms/src/cmsplugin.c
@@ -179,6 +179,8 @@ cmsBool CMSEXPORT  _cmsReadFloat32Number(cmsIOHANDLER* io, cmsFloat32Number* n)
 
         tmp = _cmsAdjustEndianess32(tmp);
         *n = *(cmsFloat32Number*) (void*) &tmp;
+        if (isnan(*n))
+            return FALSE;
     }
     return TRUE;
 }
