diff --git a/source/i18n/number_decimfmtprops.cpp b/source/i18n/number_decimfmtprops.cpp
index 6754fe19..b8994e56 100644
--- a/source/i18n/number_decimfmtprops.cpp
+++ b/source/i18n/number_decimfmtprops.cpp
@@ -15,7 +15,7 @@ using namespace icu::number::impl;
 
 namespace {
 
-char kRawDefaultProperties[sizeof(DecimalFormatProperties)];
+alignas(DecimalFormatProperties) char kRawDefaultProperties[sizeof(DecimalFormatProperties)];
 
 icu::UInitOnce gDefaultPropertiesInitOnce = U_INITONCE_INITIALIZER;
 
