diff --git a/source/i18n/decimfmt.cpp b/source/i18n/decimfmt.cpp
index a2638bb7..8ae773b7 100644
--- a/source/i18n/decimfmt.cpp
+++ b/source/i18n/decimfmt.cpp
@@ -1056,6 +1056,7 @@ UBool DecimalFormat::areSignificantDigitsUsed() const {
 }
 
 void DecimalFormat::setSignificantDigitsUsed(UBool useSignificantDigits) {
+    if (areSignificantDigitsUsed()) return;
     // These are the default values from the old implementation.
     int32_t minSig = useSignificantDigits ? 1 : -1;
     int32_t maxSig = useSignificantDigits ? 6 : -1;
