diff --git a/third_party/tlslite/tlslite/constants.py b/third_party/tlslite/tlslite/constants.py
index 84bb703..715def9 100644
--- a/third_party/tlslite/tlslite/constants.py
+++ b/third_party/tlslite/tlslite/constants.py
@@ -55,11 +55,11 @@ class ExtensionType:    # RFC 6066 / 4366
     srp = 12            # RFC 5054  
     cert_type = 9       # RFC 6091
     signed_cert_timestamps = 18  # RFC 6962
-    extended_master_secret = 23  # draft-ietf-tls-session-hash-06
+    extended_master_secret = 23  # RFC 7627
+    token_binding = 24           # draft-ietf-tokbind-negotiation
     tack = 0xF300
     supports_npn = 13172
     channel_id = 30032
-    token_binding = 30033
 
 class HashAlgorithm:
     none = 0
