diff --git a/third_party/tlslite/tlslite/tlsconnection.py b/third_party/tlslite/tlslite/tlsconnection.py
index e42b362..41aab85 100644
--- a/third_party/tlslite/tlslite/tlsconnection.py
+++ b/third_party/tlslite/tlslite/tlsconnection.py
@@ -1570,6 +1570,8 @@ class TLSConnection(TLSRecordLayer):
                 #Set the session
                 self.session = session
                     
+                self.clientRandom = clientHello.random
+                self.serverRandom = serverHello.random
                 yield None # Handshake done!
 
         #Calculate the first cipher suite intersection.
