diff --git a/third_party/tlslite/tlslite/tlsconnection.py b/third_party/tlslite/tlslite/tlsconnection.py
index 4f3ba1c..cb743fe 100644
--- a/third_party/tlslite/tlslite/tlsconnection.py
+++ b/third_party/tlslite/tlslite/tlsconnection.py
@@ -1269,6 +1269,9 @@ class TLSConnection(TLSRecordLayer):
                 return # Handshake was resumed, we're done 
             else: break
         (clientHello, cipherSuite) = result
+
+        # Save the ClientHello for external code to query.
+        self.clientHello = clientHello
         
         #If not a resumption...
 
