diff --git a/third_party/tlslite/tlslite/messages.py b/third_party/tlslite/tlslite/messages.py
index 1ce9320..ac7e563 100644
--- a/third_party/tlslite/tlslite/messages.py
+++ b/third_party/tlslite/tlslite/messages.py
@@ -223,7 +224,7 @@ class ClientHello(HandshakeMsg):
                         p2 = Parser(tokenBindingBytes)
                         ver_minor = p2.get(1)
                         ver_major = p2.get(1)
-                        if (ver_major, ver_minor) >= (0, 6):
+                        if (ver_major, ver_minor) >= (0, 13):
                             p2.startLengthCheck(1)
                             while not p2.atLengthCheck():
                                 self.tb_client_params.append(p2.get(1))
@@ -421,7 +431,7 @@ class ServerHello(HandshakeMsg):
             w2.add(4, 2)
             # version
             w2.add(0, 1)
-            w2.add(6, 1)
+            w2.add(13, 1)
             # length of params (defined as variable length <1..2^8-1>, but in
             # this context the server can only send a single value.
             w2.add(1, 1)
