diff --git a/src/libxslt/security.c b/src/libxslt/security.c
index 89de90bd3f1e..965175f79ca4 100644
--- src/libxslt/security.c
+++ src/libxslt/security.c
@@ -282,7 +282,7 @@ xsltCheckFilename (const char *path)
 #if defined(_WIN32) && !defined(__CYGWIN__)
     DWORD dwAttrs;
 
-    dwAttrs = GetFileAttributes(path);
+    dwAttrs = GetFileAttributesA(path);
     if (dwAttrs != INVALID_FILE_ATTRIBUTES) {
         if (dwAttrs & FILE_ATTRIBUTE_DIRECTORY) {
             return 2;
