// polymer-analyzer/lib/index.js
"use strict";
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This file describes the public API of the analyzer.
 *
 * Only this file and the objects reachable from its exports are considered
 * part of the stable API of the analyzer, in a semver sense.
 */
// Core objects
var analyzer_1 = require("./core/analyzer");
exports.Analyzer = analyzer_1.Analyzer;
__export(require("./model/model"));
var warning_printer_1 = require("./warning/warning-printer");
exports.WarningPrinter = warning_printer_1.WarningPrinter;
var warning_filter_1 = require("./warning/warning-filter");
exports.WarningFilter = warning_filter_1.WarningFilter;
var namespace_1 = require("./javascript/namespace");
exports.Namespace = namespace_1.Namespace;
var document_1 = require("./parser/document");
exports.ParsedDocument = document_1.ParsedDocument;
// Analysis
var generate_analysis_1 = require("./analysis-format/generate-analysis");
exports.generateAnalysis = generate_analysis_1.generateAnalysis;
exports.validateAnalysis = generate_analysis_1.validateAnalysis;
// URL Loaders and Resolvers
var fetch_url_loader_1 = require("./url-loader/fetch-url-loader");
exports.FetchUrlLoader = fetch_url_loader_1.FetchUrlLoader;
var fs_url_loader_1 = require("./url-loader/fs-url-loader");
exports.FSUrlLoader = fs_url_loader_1.FSUrlLoader;
var overlay_loader_1 = require("./url-loader/overlay-loader");
exports.InMemoryOverlayUrlLoader = overlay_loader_1.InMemoryOverlayUrlLoader;
var multi_url_loader_1 = require("./url-loader/multi-url-loader");
exports.MultiUrlLoader = multi_url_loader_1.MultiUrlLoader;
var multi_url_resolver_1 = require("./url-loader/multi-url-resolver");
exports.MultiUrlResolver = multi_url_resolver_1.MultiUrlResolver;
var package_url_resolver_1 = require("./url-loader/package-url-resolver");
exports.PackageUrlResolver = package_url_resolver_1.PackageUrlResolver;
var prefixed_url_loader_1 = require("./url-loader/prefixed-url-loader");
exports.PrefixedUrlLoader = prefixed_url_loader_1.PrefixedUrlLoader;
var redirect_resolver_1 = require("./url-loader/redirect-resolver");
exports.RedirectResolver = redirect_resolver_1.RedirectResolver;
// Polymer
var polymer_element_1 = require("./polymer/polymer-element");
exports.PolymerElement = polymer_element_1.PolymerElement;
var behavior_1 = require("./polymer/behavior");
exports.PolymerBehavior = behavior_1.Behavior;
var polymer_element_mixin_1 = require("./polymer/polymer-element-mixin");
exports.PolymerElementMixin = polymer_element_mixin_1.PolymerElementMixin;
var expression_scanner_1 = require("./polymer/expression-scanner");
exports.PolymerDatabindingExpression = expression_scanner_1.DatabindingExpression;
exports.AttributeDatabindingExpression = expression_scanner_1.AttributeDatabindingExpression;
exports.JavascriptDatabindingExpression = expression_scanner_1.JavascriptDatabindingExpression;
var dom_module_scanner_1 = require("./polymer/dom-module-scanner");
exports.DomModule = dom_module_scanner_1.DomModule;
// ParsedDocuments
var json_document_1 = require("./json/json-document");
exports.ParsedJsonDocument = json_document_1.ParsedJsonDocument;
var javascript_document_1 = require("./javascript/javascript-document");
exports.ParsedJavaScriptDocument = javascript_document_1.JavaScriptDocument;
var html_document_1 = require("./html/html-document");
exports.ParsedHtmlDocument = html_document_1.ParsedHtmlDocument;
var css_document_1 = require("./css/css-document");
exports.ParsedCssDocument = css_document_1.ParsedCssDocument;
