diff --git a/lib/constants.js b/lib/constants.js
index 6e1a557..23d7f7c 100644
--- a/lib/constants.js
+++ b/lib/constants.js
@@ -18,7 +18,7 @@ exports.default = {
     ABS_URL: /(^\/)|(^#)|(^[\w-\d]*:)/,
     URL: /url\([^)]*\)/g,
     URL_ATTR: ['href', 'src', 'action', 'style', 'assetpath'],
-    URL_TEMPLATE: '{{.*}}|\\[\\[.*\\]\\]',
+    URL_TEMPLATE: '{{.*}}|\\[\\[.*\\]\\]|\\$i18n[^{]*{[^}]*}',
     OLD_POLYMER: 'This version of polymer-bundler is not compatible with Polymer < 0.8. Please use vulcanize 0.7.x.'
 };
 //# sourceMappingURL=constants.js.map
